### ISSUE TITLE
Golang integration attempts new install even when go is installed on system.

### ISSUE THREAD

**Issue by apolo49** [pre-commit/pre-commit#3260]
### search you tried in the issue tracker

zipfile.BadZipFile: File is not a zip file, golang install

### describe your issue

When running `pre-commit run`, pre-commit fails to create go environments as it cannot download golang on my company laptop (go downloads are blocked). I have go installed on my local machine (version 1.22.5) and so go does not need to be downloaded and can instead be re-used.

When installing, because it cannot be downloaded the zipfile downloaded is likely to not be a true zip and instead be an HTML webpage stating that it cannot be downloaded, giving me the following error.

I know this is isolated to go hooks as the python pre-commit hooks work fine. (I have previously commented out the go hooks and left in the python hooks).

I think we need extra logic and/or error handling to discover if go is in the path and if it fails to download/install we should handle that case too with a better error message.

### pre-commit --version

pre-commit 3.7.1

### .pre-commit-config.yaml

```yaml
repos:
  - repo: https://github.com/gitleaks/gitleaks
    rev: v8.18.4
    hooks:
      - id: gitleaks
  - repo: https://github.com/trufflesecurity/trufflehog
    rev: v3.79.0
    hooks:
      - id: trufflehog
  - repo: https://github.com/crate-ci/typos
    rev: typos-v0.10.26
    hooks:
      - id: typos
  - repo: https://github.com/openstack/bashate
    rev: '2.1.1'
    hooks:
      - id: bashate
  - repo: https://github.com/google/yamlfmt
    rev: v0.13.0
    hooks:
      - id: yamlfmt
  - repo: https://github.com/adrienverge/yamllint
    rev: v1.35.1
    hooks:
      - id: yamllint
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v4.6.0
    hooks:
      - id: trailing-whitespace
      - id: end-of-file-fixer
      - id: check-added-large-files
      - id: check-case-conflict
      - id: detect-private-key
      - id: no-commit-to-branch
        args:
          - --branch
          - main
          - --branch
          - master
          - --branch
          - develop
      - id: check-yaml
        args:
          - --allow-multiple-documents
```

### version information

```
pre-commit version: 3.7.1
git --version: git version 2.39.2.windows.1
sys.version:
    3.12.3 (tags/v3.12.3:f6650f9, Apr  9 2024, 14:05:25) [MSC v.1938 64 bit (AMD64)]
sys.executable: C:\Users\joe.targett\AppData\Local\Programs\Python\Python312\python.exe
os.name: nt
sys.platform: win32
```

### error information

```
An unexpected error has occurred: BadZipFile: File is not a zip file
```

```
Traceback (most recent call last):
  File "C:\Users\joe.targett\AppData\Local\Programs\Pytho`n\Python312\Lib\site-packages\pre_commit\error_handler.py", line 73, in error_handler
    yield
  File "C:\Users\joe.targett\AppData\Local\Programs\Python\Python312\Lib\site-packages\pre_commit\main.py", line 417, in main
    return run(args.config, store, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\joe.targett\AppData\Local\Programs\Python\Python312\Lib\site-packages\pre_commit\commands\run.py", line 442, in run
    install_hook_envs(to_install, store)
  File "C:\Users\joe.targett\AppData\Local\Programs\Python\Python312\Lib\site-packages\pre_commit\repository.py", line 238, in install_hook_envs
    _hook_install(hook)
  File "C:\Users\joe.targett\AppData\Local\Programs\Python\Python312\Lib\site-packages\pre_commit\repository.py", line 94, in _hook_install
    lang.install_environment(
  File "C:\Users\joe.targett\AppData\Local\Programs\Python\Python312\Lib\site-packages\pre_commit\languages\golang.py", line 138, in install_environment
    _install_go(version, env_dir)
  File "C:\Users\joe.targett\AppData\Local\Programs\Python\Python312\Lib\site-packages\pre_commit\languages\golang.py", line 118, in _install_go
    with _open_archive(f) as archive:
         ^^^^^^^^^^^^^^^^
  File "C:\Users\joe.targett\AppData\Local\Programs\Python\Python312\Lib\site-packages\pre_commit\languages\golang.py", line 54, in _open_archive
    return zipfile.ZipFile(bio)
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\joe.targett\AppData\Local\Programs\Python\Python312\Lib\zipfile\__init__.py", line 1349, in __init__
    self._RealGetContents()
  File "C:\Users\joe.targett\AppData\Local\Programs\Python\Python312\Lib\zipfile\__init__.py", line 1416, in _RealGetContents
    raise BadZipFile("File is not a zip file")
zipfile.BadZipFile: File is not a zip file
```

**Comment by asottile** [comment]
please show your output, is this persistent? what's your PATH, where is `go` installed, etc.

you've given very little to go on here

**Comment by apolo49** [comment]
Go is on my user-path and is available under the command `go` on my pipeline, like any other executable on the path.

```bash
echo $env:Path
C:\Program Files\PowerShell\7;C:\Program Files (x86)\Common Files\Oracle\Java\javapath;C:\windows\system32;C:\windows;C:\windows\System32\Wbem;C:\windows\System32\WindowsPowerShell\v1.0\;C:\windows\System32\OpenSSH\;C:\Program Files (x86)\Webex\Webex\Applications;C:\Program Files\Git\cmd;C:\Program Files\dotnet\;C:\Program Files\PowerShell\7\;C:\Users\joe.targett\AppData\Local\Microsoft\dotnet\;C:\Users\joe.targett\AppData\Local\Programs\Python\Python312\Scripts\;C:\Users\joe.targett\AppData\Local\Programs\Python\Python312\;C:\Users\joe.targett\AppData\Local\Programs\Python\Launcher\;C:\Users\joe.targett\AppData\Local\Microsoft\WindowsApps;C:\ProgramData\chocoportable\bin;C:\Users\joe.targett\AppData\Local\Programs\Microsoft VS Code\bin;C:\Users\joe.targett\AppData\Local\Microsoft\WindowsApps\go\bin
```

It is installed in an unusual place, but as it is on my path, this should not be a problem and should be taken into account.

**Comment by asottile** [comment]
executables in the home directory are explicitly ignored by pre-commit as they are easily mutable breaking the cache in surprising ways

**Comment by asottile** [comment]
(you can override this explicitly by selection `language_version: system` -- but it is at your own risk)

**Comment by asottile** [comment]
this doesn't explain the other part though -- would you be able to drop a `breakpoint()` and see what it's trying to download?

**Comment by apolo49** [comment]
> executables in the home directory are explicitly ignored by pre-commit as they are easily mutable breaking the cache in surprising ways

I am sorry, I am confused by this, this executable isn't technically located in the home directory, but instead in a subdirectory. Are you saying user-installed programs are ignored (they need to be installed for the entire system)? This could potentially alienate a large userbase who do not have admin permissions to install to system (i.e. me).

Where would you like the breakpoint?

**Comment by asottile** [comment]
> I am sorry, I am confused by this, this executable isn't technically located in the home directory, but instead in a subdirectory

not sure why you're confused by this -- a directory in the home directory is still in the home directory.  just because it has a directory component in it does not mean it is not inside the home directory.

> Are you saying user-installed programs are ignored (they need to be installed for the entire system)?

yes and no, they are not considered for the *default* language version but they can still be used as above

>  This could potentially alienate a large userbase who do not have admin permissions to install to system (i.e. me).

it has always been like this and generally has not done what you're saying because pre-commit will successfully install the language for them.  you're hitting some weird case where that's not working -- perhaps due to corporate firewall, etc.

> Where would you like the breakpoint?

where it would be useful!  figure out what's being downloaded and why it's not what it says it is

**Comment by apolo49** [comment]
I can confirm it is trying to download the following file: `https://dl.google.com/go/go1.22.5.windows-amd64.zip`

The response is the zip file and I can open the temp file using 7-zip successfully. Not sure why it isn't working.

**Comment by apolo49** [comment]
It seems to install to a symlink? Not sure if this would cause an error (`f.name` line 117 in golang.py):
`C:\Users\JOE~1.TAR\AppData\Local\Temp\<temp-file-name>`
I think this should instead point to where explorer takes me `C:\Users\joe.targett\AppData\Local\Temp\<temp-file-name>`

**Comment by asottile** [comment]
that's a windows "8.3" path, nothing special there and not the problem (it's partially caused by your user name having a `.` in it -- but that's off topic here)

**Comment by asottile** [comment]
if you try and open that file with python what happens?

it works fine for me fwiw:

```pycon
>>> zipfile.ZipFile(open('go1.22.5.windows-amd64.zip', 'rb'))
<zipfile.ZipFile file=<_io.BufferedReader name='go1.22.5.windows-amd64.zip'> mode='r'>
```

**Comment by apolo49** [comment]
```python
>>> import zipfile
>>> zipfile.ZipFile(open('test', 'rb'))
<zipfile.ZipFile file=<_io.BufferedReader name='test'> mode='r'>
```

Downloaded as `test` using cURL.
```bash
curl -k -H 'Accept-Content: application/zip' -H 'Accept-Encoding: gzip, deflate, br, zstd' 
https://dl.google.com/go/go1.22.5.windows-amd64.zip -o $LOCALAPPDATA/test
```

**Comment by apolo49** [comment]
Even using the downloaded copy from pre-commit it opens:
```python
>>> zipfile.ZipFile(open('tmpa1n_o9ei', 'rb'))
<zipfile.ZipFile file=<_io.BufferedReader name='tmpa1n_o9ei'> mode='r'>
```

code segment in pre-commit to see changes
```python
def _install_go(version: str, dest: str) -> None:
    try:
        resp = urllib.request.urlopen(_get_url(version))
    except urllib.error.HTTPError as e:  # pragma: no cover
        if e.code == 404:
            raise ValueError(
                f'Could not find a version matching your system requirements '
                f'(os={platform.system().lower()}; arch={_ARCH})',
            ) from e
        else:
            raise
    else:
        with tempfile.TemporaryFile(delete=False) as f:
            shutil.copyfileobj(resp, f)
            f.seek(0)

            with _open_archive(f) as archive:
                archive.extractall(dest)
        shutil.move(os.path.join(dest, 'go'), os.path.join(dest, '.go'))
```

**Comment by apolo49** [comment]
All I added was `delete=False`

**Comment by asottile** [comment]
have you retried just running the pre-commit parts? possible it was just a blip?

**Comment by apolo49** [comment]
I've been trying for days... I don't think it's a blip (not impossible though I guess.) I'll try again

**Comment by apolo49** [comment]
That's odd, it just worked? I'll let you know if this comes up again, but this is now super odd!

**Comment by asottile** [comment]
shrugs it happens -- the error seemed like something where google gave you a 200 with an error page 🤷

**Comment by faymard** [comment]
Hi, I found this issue while investigating an issue caused by Nodejs being installed in the home directory. Please tell me if I need to open a new one.

> executables in the home directory are explicitly ignored by pre-commit as they are easily mutable breaking the cache in surprising ways

@asottile Agreed. But we stumbled across a case that is very difficult to find a solution for. Recently NodeJS released v23.0.0, and it broke the path library, breaking **npm pack** (https://github.com/nodejs/node/issues/55410). As a consequence, pre-commits based on NodeJS would break because of this if :
* Node was installed in the home directory
* Node was not installed

This is obviously not a fault of pre-commit tool.

We tried to set `default_language_version:` in the pre-commit-config file, which after all could be seen as a good practice. It resolved hook execution on our workstations (linux amd64). This looks good, but our CI pre-commit image on which we installed Node in the system would now break (alpine arm64, so musl-based and not compatible with official Node builds). 

Would it be possible to have an opt-in configuration property somewhere to authorize home-directory executables ? This way, our workstations could use their own Node binaries while our CI images would still be shipped with system-level Node installs.

I would be interested in making that contribution but I wanted to have your point of view on this proposal before coding anything.

Regards and thank you for this widely-used tool

**Comment by asottile** [comment]
no means no.  please stop trying to relitigate
